src/stirling/obj_tools/go_syms.cc

```diff
@@ -44,48 +47,101 @@ bool IsGoExecutable(ElfReader* elf_reader) {
   return elf_reader->SearchTheOnlySymbol(obj_tools::kGoBuildVersionSymbol).ok();
 }
 
-// TODO(ddelnano): Between Go 1.20.2 and 1.20.4 our build version detection started failing.
-// Our version detection's assumptions is very different from how Go does this internally
-// and needs a signficant rehaul. That is being tracked in
-// https://github.com/pixie-io/pixie/issues/1318 but in the meantime optimisitcally read the
-// runtime.buildVersion.str before following our previous heuristic.
-StatusOr<std::string> ReadBuildVersionDirect(ElfReader* elf_reader) {
-  auto str_symbol_status = elf_reader->SearchTheOnlySymbol(kGoBuildVersionStrSymbol);
-  if (!str_symbol_status.ok()) {
-    return error::NotFound("Unable to find runtime.buildVersion.str");
-  }
-  auto str_symbol = str_symbol_status.ValueOrDie();
-  PX_ASSIGN_OR_RETURN(auto symbol_bytecode, elf_reader->SymbolByteCode(".rodata", str_symbol));
-  return std::string(reinterpret_cast<const char*>(symbol_bytecode.data()),
-                     symbol_bytecode.size() - 1);
+constexpr std::string_view kGoBuildInfoSection = ".go.buildinfo";
+// kGoBuildInfoMagic corresponds to "\xff Go buildinf:"
+// https://github.com/golang/go/blob/1dbbafc70fd3e2c284469ab3e0936c1bb56129f6/src/debug/buildinfo/buildinfo.go#L49
+std::string_view kGoBuildInfoMagic =
+    CreateStringView<char>("\xff\x20\x47\x6f\x20\x62\x75\x69\x6c\x64\x69\x6e\x66\x3a");
+
+// Reads a Go string encoded within a buildinfo header. This function is meant to provide the same
+// functionality as
+// https://github.com/golang/go/blob/master/src/debug/buildinfo/buildinfo.go#L244C37-L244C44
+StatusOr<std::string> ReadGoString(ElfReader* elf_reader, uint64_t ptr_size, uint64_t ptr_addr,
+                                   read_ptr_func_t read_ptr) {
+  PX_ASSIGN_OR_RETURN(u8string_view data_addr, elf_reader->BinaryByteCode(ptr_addr, ptr_size));
+  PX_ASSIGN_OR_RETURN(u8string_view data_len,
+                      elf_reader->BinaryByteCode(ptr_addr + ptr_size, ptr_size));
+
+  PX_ASSIGN_OR_RETURN(uint64_t vaddr_offset, elf_reader->GetVirtualAddrAtOffsetZero());
+  ptr_addr = read_ptr(data_addr) - vaddr_offset;
+  uint64_t str_length = read_ptr(data_len);
+
+  PX_ASSIGN_OR_RETURN(std::string_view go_version_bytecode,
+                      elf_reader->BinaryByteCode<char>(ptr_addr, str_length));
+  return std::string(go_version_bytecode);
 }
 
+// Reads the buildinfo header embedded in the .go.buildinfo ELF section in order to determine the go
+// toolchain version. This function emulates what the go version cli performs as seen
+// https://github.com/golang/go/blob/cb7a091d729eab75ccfdaeba5a0605f05addf422/src/debug/buildinfo/buildinfo.go#L151-L221
 StatusOr<std::string> ReadBuildVersion(ElfReader* elf_reader) {
```

Comment:
It's because many of the uprobe deployment helpers need to perform some form of ELF parsing. At the start of each UProbe deployment function ([Go](https://github.com/pixie-io/pixie/blob/6e0c199ea00f55ff7ef7de2e063d0f342526b6c4/src/stirling/source_connectors/socket_tracer/uprobe_manager.cc#L850), [NodeJS](https://github.com/pixie-io/pixie/blob/6e0c199ea00f55ff7ef7de2e063d0f342526b6c4/src/stirling/source_connectors/socket_tracer/uprobe_manager.cc#L513), [OpenSSL](https://github.com/pixie-io/pixie/blob/6e0c199ea00f55ff7ef7de2e063d0f342526b6c4/src/stirling/source_connectors/socket_tracer/uprobe_manager.cc#L464)), we create a shared ELFReader and pass it to all of the functions that need it.